news/_posts/2019-09-11-prestashop-in-2019-and-beyond-part-2-pain-points.md

```diff
@@ -0,0 +1,471 @@
+- **Want to refactor that class to make it testable?** _No can do, that would require changing its signature_.
+- **Want to upgrade that dependency to benefit from cool new features?** _Not if it's a major version upgrade_.
+- :'(
+
+### Everything is public
+
+As if making every class part of our API was not enough of a problem, a lot of legacy core classes suffer from _public_-itis. It means that classes have lots of public methods and properties that needn't be, just for convenience's sake (violating the [encapsulation principle](https://en.wikipedia.org/wiki/Encapsulation_(computer_programming))).
+
+This is particularly troublesome in PrestaShop, because once a class member has become public—even if it was a mistake—it becomes part of the public API as soon as it's released... and therefore cannot be refactored until the next major version!
+
+It gets _even worse_ because of the [overrides system][overrides-system], which allows developers to replace any legacy core class with a custom one (violating the [Open–Closed principle](https://en.wikipedia.org/wiki/Open%E2%80%93closed_principle), by the way). Since the overrides system is based on inheritance, and inheritance grants overrides access to protected class members of the extended class, and any accessible code is considered part of the API... then **protected members of legacy core classes can be considered part of the public API as well**, because they are accessible. 
+
+Having no clearly-defined contracts means we can't change anything without introducing a breaking change, which severely limits the options for improving PrestaShop.
+
+## 1.7 was released too soon
+
+PrestaShop 1.7.0 was a very ambitious release. It featured a [new architecture](/news/new-architecture-1-6-1-0/), a [new framework (Symfony)](/news/prestashop-1-7-and-symfony/), a new Back Office theme (later based on our [UI kit](/news/PrestaShop-UI-Kit/)), a [new default Front Office theme](/news/new-theme-1-7-introduction/) and [Starter Theme](/starter-theme-kickoff/), a [new Product page experience](/news/product-page-evolution/), a [new Modules page experience](/module-page-awakens/), a [new translation system](/news/new-translation-system-prestashop-17/) and [use of CLDR](/cldr-composer-jshint/), among many other features. 
+
+That's an impressive list. Unfortunately, that objective was proven to be too ambitious, and the result was an underwhelming initial release.
+
+Before continuing, it's important to say that following this release and between 2016 and 2018, most of the original team that developed PrestaShop 1.7 left the company or was reassigned. In 2017, a new CTO (Aurélien Pelletier) and I arrived at PrestaShop. We agreed on a new long-term vision that would point the project in the right direction, and started building a new team to make it happen. Early 2018, with the support of the newly appointed PrestaShop CEO, Alexandre Eruimy, the team started growing and the project started picking up speed again.
```

Comment:
💪

Reply:
I needed to say this because I can't justify what's been done in the past, and I don't think the new team deserves the blame.